src/Sulu/Bundle/MediaBundle/Content/Types/SingleMediaSelection.php

```diff
@@ -123,4 +126,27 @@ protected function decodeValue($value)
 
         return \json_decode($value, true);
     }
+
+    public function mapPropertyMetadata(PropertyMetadata $propertyMetadata): ObjectMetadata
+    {
+        $mandatory = $propertyMetadata->isRequired();
+
+        $jsonSchema = [
+            'type' => 'object',
```

Comment:
The fact that you have to pass `object` as a type here, although you are generating a `ObjectMetadata` below might lead to some confusing situations... What if we create some like that:

```php
return new ObjectMetadata($propertyMetadata->getName(), $mandatory, ['type' => 'string']);
```

That's why I think it might be confusing to build it this way. Would it be good enough for the use cases we have to build it like this:

```php
new ObjectMetadata(
    $propertyMetadata->getName(),
    $mandatory,
    [
        new PropertyMetadata('id', true, 'number'),
        new PropertyMetadata('displayOption', false, 'string'),
    ]
);
```

In that example I assume that we are adding a third argument to the `PropertyMetadata` for the type. Instead we could also add something like a `StringMetadata` and `NumberMetadata`.

Then it would be consistent IMO. I know that this might be a little bit cumbersome if somebody is a JSON schema pro, and they just want to pass an array as schema. I would totally understand if we want to support that better, but then I would deprecate all of the classes (including `SchemaMetadata`, `PropertyMetadata`, ...) and only pass one - potentially huge - array instead.

My suggestion would be to try it the way I have mentioned above, and if that turns out to be too cumbersome in the future, we could still deprecate all these classes, and build it in a way that just an array is passed (would probably need a service then, that would help with merging schemas etc., basically replacing the logic that is happening in the `SchemaMetadata` classes).

Reply:
Your suggestions works very well for simple json schemas ... But if you have a short look at https://github.com/sulu/sulu/pull/5640/files#diff-6f4f6054791755b1732dec2b54938638b8a7dd63b892bcf3b0ef7d63e1358946R200-R215, it already gets a little bit more complicated. There is already an `ArrayMetadata`, but that would have to be fully modified to represent e.g. the schema needed for the `ids` property. I don't even think, that this can be done without a bc break

Reply:
Are you talking about the `uniqueItems` and `minItems` properties? These can be optional, then it's not a BC break, or am I missing something obvious :thinking:

Reply:
The `ArrayMetadata` can't handle scalar arrays as far as I see

Reply:
Also, we would have to add a parameter to `ArrayMetadata` and to all other `Metadata`s for every json schema keyword we want to support, so this would be a big limitation

Reply:
Wouldn't something like that work:

```php
new ObjectMetadata(
    $propertyMetadata->getName(),
    $mandatory,
    [
        new ArrayMetadata('ids', true, new SchemaMetadata([], [], [], 'string')),
        new PropertyMetadata('displayOption', false, 'string'),
    ]
);
```

I know that we would have to add a new parameter to the `SchemaMetadata`, but that could also easily be hidden in factory methods (or we just leave it that way for now). That would need a small change, but I still don't think that it would be a breaking one.

Regarding the limitation: That's correct, but I am not sure if that will be a real deal breaker. Especially since in order to properly handle new keywords we sometimes might have to add some stuff in the `Form` javascript implementation.

As said, that's why I would at least try to implement it that way. My guess is that mixing classes for some stuff and plain arrays in other cases would cause more confusion that it helps. With this solution we would keep this consistent for now, and I am happy to change it later to a pure array variant (deprecating all the classes in the `Sulu\Bundle\AdminBundle\Metadata\SchemaMetadata`) if it really turns out to be cumbersome. But I think it would be a pity to lose consistency now.

Reply:
ok 👍